Fix redirect when deleting current conversation

PVince81 opened the pull request:
When the current conversation was deleted while in a call, the redirect
now targets the not-found page instead of the one about duplicate
session.

Added event bus handler for the not found page which now also removes
the session store's conversation token.

Fixes https://github.com/nextcloud/spreed/issues/4552

PVince81 commented:
/backport to stable21

nickvergessen commented on src/App.vue:
```diff
@@ -233,11 +234,23 @@ export default {
+			this.$nextTick(() => {
+				// Need to delay until next tick, otherwise the PreventUnload is still being triggered
+				// Putting the window in front with the warning and irritating the user
+				window.location = generateUrl('/apps/spreed/not-found')
```
Why this over `this.$router.push('/apps/spreed/not-found')` ?

nickvergessen replied:
Also this doesn't seem to work. For me the "Leave call" popup is still shown (without HPB)

PVince81 replied:
ah, I copied this from mixins/duplicateSessionHandler which also doesn't use router push.
I'll check if it works with that.

PVince81 replied:
> Also this doesn't seem to work. For me the "Leave call" popup is still shown (without HPB)

It works if the deletion is done outside the current browser window as per original reproduction steps.

But indeed if you delete it within the same window you get the warnings. I'll try to add special handling for that scenario.

PVince81 replied:
I've tried using router.push() but it's triggering the "you are leaving call" dialog.
When adding a flag to inhibit it, it will display the not found page but with a lot of JS errors...

I don't think it's worth trying to protect against this at this stage (but worth as part of improving the not found page separately).

I'll leave window.location here then. I'm sure this must have been the reason why it was used also for the duplicate session case.

nickvergessen replied:
Well signaling.js is not Vue, so it can't use router.push()
It smells so trappy to change this. There must be a reason we did it like this and not do a page reload all the time.

PVince81 replied:
> It smells so trappy to change this.

I'd appreciate a more diplomatic tone...

I'll have another look and compare with the original behavior and see why the errors I saw didn't happen there.

PVince81 commented:
@nickvergessen I've pushed a fix to prevent leaving or deleting a conversation while in a call. This will prevent the can of worms of trying to handle warning dialogs, etc.

As a bonus I also added a "close that freaky popup already when I click".

Regarding router push, see my comment above. It's also a can of worms.

nickvergessen commented on src/components/LeftSidebar/ConversationsList/Conversation.vue:
```diff
@@ -84,11 +84,15 @@
 
 			<ActionButton v-if="canLeaveConversation"
+				:close-after-click="true"
+				:disabled="item.hasCall"
```
* So I create a conversation
* Invite the full company
* Start a call
* No one can leave it
* muhahaha

Maybe only when you are part of the call?

PVince81 replied:
I knew something was fishy... Couldn't be that easy.

To detect "is in call" need to bring in two mixins :-/

I'll see if I can still solve this differently and inhibit the dialog while the redirect is in progress.

PVince81 replied:
I can try moving this computed property to the store. Store getters are supposed to be like global computed properties and we've needed this one in enough places already.

PVince81 replied:
ok, as discussed this is a can of worms

PVince81 commented:
I've changed the approach. The new approach doesn't refactor the old code.

Since we'll need this kind of handling outside of App.vue, I've put the code into the mixin duplicateSessionHandler which I renamed to sessionIssueHandler.

Cases to cover and test:
- [ ] deleting session from outside
   - [x] regular call
   - [ ] sidebar call
   - [ ] video verification
   - [ ] guest call
- [ ] deleting from left sidebar while in the same call: todo, see https://github.com/nextcloud/spreed/pull/5082#discussion_r573885095

PVince81 commented:
looks like we were redirecting to talk's "not found" page already in a sidebar call, despite being in the files app :-/
I'll make it the same now, we can try and find a cleaner solution for that later... maybe need to embed a "not found" in the sidebar tab and make sure it's recoverable

PVince81 commented:
it seems that when deleting the current conversation there's already another handler that sets the route because it reaches the sessionIssueHandler, this is what triggers that dialog.
I might need a second mechanism to inhibit the dialog in this code path.

PVince81 commented:
Aha, finally found something: when setting the route to not-found it doesn't automatically clear the token, this is why I've seen a lot of JS errors because the right sidebar is still present with its chat tab.

PVince81 commented:
After some research and tests I realized the following:
- we can push the "notfound" route instead of doing a full page reload:
   - the JS errors I saw before were only related to the fact that we forgot to clear the current token in some places
   - I've applied this now to this PR and it looks smooth when getting kicked out...
   - instead of a semi-global attribute for saying "please don't warn about leaving", we can use a route param when using named route, see https://router.vuejs.org/guide/essentials/navigation.html
- duplicate session handling must trigger a page reload to avoid side effects
   - in my local tests if I simply push the duplicate session route, it interferes with the other tab where I click "Join now" and that other tab gets a "not found"
   - there's probably something not cleant up properly. Considering that we'll [get rid of dupe sessions soonish](https://github.com/nextcloud/spreed/issues/1587) I think it's not worth the effort fixing this.

@nickvergessen please re-review

PVince81 commented:
@nickvergessen any concerns against merging this ?

I've noticed this might tie in with my discovery in https://github.com/nextcloud/spreed/issues/4670#issuecomment-781474242 where we could also make the "go to not found" event only work if the token we're leaving is also the one on the page.
This could be achieved by passing the token from the 404 error to the event. Then the event handler compares the passed token with the one in store. If it's the same, redirect. Else: ignore.

nickvergessen commented:
It's a bit weird to be asked whether you want to leave the conversation. You have no chance to decide this. The conversation is gone, so it should fully terminate already, and not only show the correct "The conversation does not exist - Join a conversation or start a new one! " at the end. 
![Bildschirmfoto von 2021-03-01 11-34-19](https://user-images.githubusercontent.com/213943/109485537-1add9b00-7a82-11eb-90d7-ef9637607e9b.png)

Left video (the moderator in the other tab) froze already.

For the moderator I see our custom "leave call" dialog for some split second:

![Peek 2021-03-01 11-37](https://user-images.githubusercontent.com/213943/109485937-858ed680-7a82-11eb-8ea2-0625a0105cbd.gif)

PVince81 commented:
When I tested in chromium back then I didn't see the browser's native dialog, because now we're using the router, it should switch you to the "not found" page without a page refresh. Strange that it didn't work.

I'll recheck and test with Firefox then.

PVince81 commented:
With the following I did not see any browser dialog:

1. Login as admin in Chromium
2. Create a conversation and invite user "alice"
3. Start a call
4. In Firefox login as "alice"
5. Join the conversation
6. Back in Chromium, open the left sidebar and delete the conversation

=> in the admin browser I land on the welcome page
=> in alice's browser I land on the "not found" page

Not sure if you used different steps or if it's a cache issue ?

PVince81 commented:
/backport to stable21.1

nickvergessen commented:
So I'm currently on the "Not found" page with Ivan in a call. We still hear each others as the webrtc stuff all didn't get canceled, but we can't change mic/camera/screenshare status anymore.

So I propose to go back to the redirect at least while being in a call.

PVince81 commented:
> So I'm currently on the "Not found" page with Ivan in a call. We still hear each others as the webrtc stuff all didn't get canceled, but we can't change mic/camera/screenshare status anymore.
> 
> So I propose to go back to the redirect at least while being in a call.

PR here: https://github.com/nextcloud/spreed/pull/5595
